azavea/climate-change-lab
# Update project deployment tooling, add docker-compose development environment

## Overview

This PR brings this project into alignment with our standard setup by updating the deployment tooling to use Terraform and AWS instead of the `s3_website` gem, and replacing the Vagrant development environment with a docker-compose based setup. 

### Notable Changes
- [x] Repace `npm` package manager with `yarn`
- [x] Replace Vagrant setup with docker-compose
- [x] Replace `s3_website` deployment with Terraform/AWS CLI 
- [x] Add docker-based deployment tooling


### Demo

~Optional. Screenshots, `curl` examples, etc.~


### Notes
- I couldn't verify that `package-lock.json` is exactly the same as yarn.lock. However, I created yarn.lock using the following steps:
```bash
$ npm install # create node_modules from package.json
$ yarn import # create yarn.lock from the contents of the node_modules folder
```


- Initially, dockerized Karma tests were failing in CI (Ubuntu 16.04) with the following error:
```bash
ChromeHeadless stderr: Failed to move to new namespace: PID namespaces supported, Network namespace supported, but failed: errno = Operation not permitted
```
I was able to get things working by creating a Karma custom launcher that passes `--no-sandbox` to Chomium. The custom launcher overrode the default launcher flags, so I had to include `--disable-gpu`, `--headless`, and `--remote-debugging-port` as well. 

## Testing Instructions

 * Rebuild the `app` VM, ensure the process completes successfully.
 * Run `script/server`, make sure the site is visible at http://localhost:4200
 * See [Jenkins output](http://urbanappsci.internal.azavea.com/job/azavea/job/climate-change-lab/job/feature%252Ftnation%252Fupdate-deployment-tooling/17/console)
 * See https://lab.staging.climate.azavea.com

Closes #223
